nibabel/brainvoyager/bv.py

```diff
@@ -0,0 +1,957 @@
+
+        # compute the translation
+        fcc = np.array(self.get_framing_cube()) / 2  # center of framing cube
+        bbc = np.array(self.get_bbox_center())  # center of bounding box
+        tra = np.dot((bbc - fcc), rot)
+
+        # assemble
+        M = np.eye(4, 4)
+        M[0:3, 0:3] = rot
+        M[0:3, 3] = tra.T
+
+        return M
+
+    get_best_affine = get_base_affine
+
+    get_default_affine = get_base_affine
+
+    get_affine = get_base_affine
+
+    def _guess_framing_cube(self):
+        """Guess the dimensions of the framing cube.
+
+        Guess the dimensions of the framing cube that constitutes the
+        coordinate system boundaries for the bounding box.
+
+        For most BV file formats this need to be guessed from
+        x_end, y_end, and z_end in the header.
+        """
+        # then start guessing...
+        hdr = self._hdr_dict
+        # get the ends of the bounding box (highest values in each dimension)
+        x = hdr['x_end']
+        y = hdr['y_end']
+        z = hdr['z_end']
+
+        # compare with possible framing cubes
+        for fc in [256, 384, 512, 768, 1024]:
+            if any([d > fc for d in (x, y, z)]):
+                continue
+            else:
+                return fc, fc, fc
+
+    def get_framing_cube(self):
+        """Get the dimensions of the framing cube.
+
+        Get the dimensions of the framing cube that constitutes the
+        coordinate system boundaries for the bounding box.
+        For most BV file formats this need to be guessed from
+        x_end, y_end, and z_end in the header.
+        """
+        return self._framing_cube
+
+    def set_framing_cube(self, fc):
+        """Set the dimensions of the framing cube.
+
+        Set the dimensions of the framing cube that constitutes the
+        coordinate system boundaries for the bounding box
+        For most BV file formats this need to be guessed from
+        x_end, y_end, and z_end in the header.
+        Use this if you know about the framing cube for the BV file.
+        """
+        self._framing_cube = fc
+
+    def get_bbox_center(self):
+        """Get the center coordinate of the bounding box.
+
+        Get the center coordinate of the bounding box with respect to the
+        framing cube.
+        """
+        hdr = self._hdr_dict
+        x = (hdr['x_start'] +
+             ((hdr['x_end'] - hdr['x_start']) / 2))
+        y = (hdr['y_start'] +
+             ((hdr['y_end'] - hdr['y_start']) / 2))
+        z = (hdr['z_start'] +
+             ((hdr['z_end'] - hdr['z_start']) / 2))
+        return z, y, x
+
+    def get_zooms(self):
+        shape = self.get_data_shape()
+        return tuple(float(self._hdr_dict['resolution'])
+                     for d in shape[0:3])
+
+    def set_zooms(self, zooms):
+        if type(zooms) == int:
+            self._hdr_dict['resolution'] = zooms
+        else:
+            if any([zooms[i] != zooms[i + 1] for i in range(len(zooms) - 1)]):
```

Comment:
In file formats with 4 dimensions (implemented are VTC and VMP) the fourth dimension is handled quite specifically. In VTCs the time dimensions is expressed in the TR header field. In VMPs the fourth dimensions is the number of submaps. In MSK files on the other hand, only three dimensions can exist. I did not come up with a smart way of generalising this.